WhateverGreen/kern_rad.cpp

```diff
@@ -183,7 +192,108 @@ void RAD::processKernel(KernelPatcher &patcher, DeviceInfo *info) {
 	}
 }
 
+uint32_t RAD::getMaxBrightnessFromDisplay() {
+	OSDictionary * matching = IOService::serviceMatching("AppleBacklightDisplay");
+	if (matching) {
+		OSIterator *iter = IOService::getMatchingServices(matching);
+		if (iter) {
+			IORegistryEntry* disp = OSDynamicCast(IORegistryEntry, iter->getNextObject());
+			if (disp) {
+				OSDictionary* iodispparm = OSDynamicCast(OSDictionary, disp->getProperty("IODisplayParameters"));
+				if (iodispparm) {
+					OSDictionary* linearbri = OSDynamicCast(OSDictionary, iodispparm->getObject("linear-brightness"));
+					if (linearbri) {
+						OSNumber* maxbri = OSDynamicCast(OSNumber, linearbri->getObject("max"));
+						callbackRAD->maxPwmBacklightLvl = maxbri->unsigned32BitValue();
+						SYSLOG("igfx", "getMaxBrightnessFromDisplay get max brightness: 0x%x", callbackRAD->maxPwmBacklightLvl);
+					} else {
+						SYSLOG("igfx", "getMaxBrightnessFromDisplay null linearbri");
+					}
+				} else {
+					SYSLOG("igfx", "getMaxBrightnessFromDisplay null iodispparm");
+				}
+			} else {
+				SYSLOG("igfx", "getMaxBrightnessFromDisplay null disp");
+			}
+			
+			iter->release();
+		} else {
+			SYSLOG("igfx", "getMaxBrightnessFromDisplay null iter");
+		}
+		
+		matching->release();
+	} else {
+		SYSLOG("igfx", "getMaxBrightnessFromDisplay null matching");
+	}
+	
+	return 0;
+}
+
+bool RAD::wrapDcLinkSetBacklightLevel(const void *dc_link, uint32_t backlight_pwm_u16_16, uint32_t frame_ramp) {
+	return FunctionCast(wrapDcLinkSetBacklightLevel, callbackRAD->orgDcLinkSetBacklightLevel)(dc_link, backlight_pwm_u16_16, frame_ramp);
+}
+
+char RAD::wrapDce110EdpBacklightControl(void *that, int64_t on_or_off, int64_t a3, int64_t a4, int64_t a5, int64_t a6) {
+	callbackRAD->dcLinkPtr = that; // save the dc link ptr for later backlight control
+	callbackRAD->getMaxBrightnessFromDisplay(); // read max brightness value from IOReg
+	return FunctionCast(wrapDce110EdpBacklightControl, callbackRAD->orgDce110EdpBacklightControl)(that, on_or_off, a3, a4, a5, a6);
+}
+
+IOReturn RAD::wrapAMDRadeonX6000AmdRadeonFramebufferSetAttribute(IOService *framebuffer, IOIndex connectIndex, IOSelect attribute, uintptr_t value) {
+	IOReturn ret = FunctionCast(wrapAMDRadeonX6000AmdRadeonFramebufferSetAttribute, callbackRAD->orgAMDRadeonX6000AmdRadeonFramebufferSetAttribute)(framebuffer, connectIndex, attribute, value);
+	if (attribute == (UInt32)'bklt') {
+		if (callbackRAD->maxPwmBacklightLvl != 0) {
+			// set the backlight of AMD navi10 driver
+			callbackRAD->curPwmBacklightLvl = value;
+			double btlper = (double)callbackRAD->curPwmBacklightLvl / callbackRAD->maxPwmBacklightLvl;
+			if (btlper < 0) {
+				btlper = 0;
+			}
+			if (btlper > 1.0) {
+				btlper = 1.0;
+			}
+			int pwmval = (int)(btlper * 0xFF) << 8;
+			if (pwmval >= 0xFF00) {
```

Comment:
This code should compare for equality exclusively if you use integer arithmetics as btlper cannot be over 1. It is not clear to me why the conversion needs to happen even with this correction, however.

Can you:
- explain this with a comment;
- do the conversion in the if assignment above after using integer maths